### PR TITLE
WIP: Concurrency

### PR DESCRIPTION
**Pull request by lalexgap**
This is work to enable concurrency for the engine and to get the integration tests passing.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ type Client struct {
 	engine              engine.Engine // The core business logic of the client
```
Comment (andrewgordstewart) [inline client/client.go:19]: @lalexgap Can this commit (1a151dd263b63b9449949599acba63f4da258899) be turned into a standalone PR?

Comment (lalexgap) [reply client/client.go:19]: Yup, I think it could be. Although I'm not super happy about having logic sprinkled everywhere to refetch the objective.

File: client/engine/channellocker.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+// NewChannelLocker returns a new ChannelLocker
+func NewChannelLocker() *ChannelLocker {
+	return &ChannelLocker{
+		channelLocks: sync.Map{},
+	}
+}
+
+// Lock acquires a locks on the given channels.
+func (l *ChannelLocker) Lock(channelIds []types.Destination) {
```
Comment (andrewgordstewart) [inline client/engine/channellocker.go:25]: I thought that `Lock` would _always_ be called with exactly two channels. Does it make sense to reflect that in the signature? This might also simplify the implementation.

Comment (lalexgap) [reply client/engine/channellocker.go:25]: It depends on what role the participant is playing in a virtual channel. If the a participant is a hub they will have two ledger channels, otherwise they'll just have one ledger channel.

In our current implementation we also need to lock the virtual channel to prevent concurrently process messages for the same objective at the same time. This means we're locking either 2 or 3 channels when locking for a virtual fund objective.

**Comment by lalexgap** [comment]
Closing this in favour of #374
